keps/sig-auth/3299-kms-v2-improvements/README.md

```diff
@@ -0,0 +1,524 @@
+    - metadata: structured data, can contain the encrypted local KEK, can be used for debugging, recovery, opaque to API server, stored unencrypted, etc. Validation similar to how K8s labels are validated today. Labels have good size limits and restrictions today.
+    - A status request and response periodically (order of minutes) returns `version`, `healthz`, and `current_key_id`
+    - The `current_key_id` in status can be used on decrypt operations to compare and validate the key ID stored in the DEK cache and the latest `EncryptResponse` `current_key_id` to detect if an object is stale in terms of storage migration
+    - Generate a new UID for each envelope operation in kube-apiserver
+    - Add a new UID field to `EncryptRequest` and `DecryptRequest`
+- Add support for hot reload of the `EncryptionConfiguration`:
+    - Watch on the `EncryptionConfiguration`
+    - When changes are detected, process the `EncryptionConfiguration` resource, and add new transformers and update existing ones atomically.
+    - If there is an issue with creating or updating any of the transformers, retain the current configuration in the kube-apiserver and generate an error in logs.
+- Enable fully automated rotation for `latest` key in KMS:
+    > NOTE: Prerequisite: `EncryptionConfiguration` is set up to always use the `latest` key version in KMS and the values can be interpreted dynamically at runtime by the KMS plugin to hot reload the current write key. Rotation process sequence:
+    - record initial key ID across all API servers (this could be recorded in the `StorageVersionStatus` as a new field)
+    - cause key rotation in KMS (user action in the remote KMS)
+    - observe the change across the stack (wait for convergence of `StorageVersionStatus`)
+    - storage migration (run storage migrator)
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+`EncryptionConfiguration` will be expanded to support the new v2 API:
+
+```diff
+​​diff --git a/staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go b/staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go
+index d7d68d2584d..84c1fa6546f 100644
+--- a/staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go
++++ b/staging/src/k8s.io/apiserver/pkg/apis/config/v1/types.go
+@@ -51,7 +51,8 @@ type ProviderConfiguration struct {
+	 // identity is the (empty) configuration for the identity transformer.
+	 Identity *IdentityConfiguration `json:"identity,omitempty"`
+	 // kms contains the name, cache size and path to configuration file for a KMS based envelope transformer.
+-    KMS *KMSConfiguration `json:"kms,omitempty"`
++    KMS   *KMSConfiguration   `json:"kms,omitempty"`
++    KMSv2 *KMSv2Configuration `json:"kmsv2,omitempty"`
```

Comment:
I'm not sure what we should do here but I think I prefer adding an API version to KMSConfiguration rather than introducing a new configuration object. If start embedding kmsv2 in field names, then we are stuck with it forever but if CacheSize and Timeout don't make sense at some point in the future we can ignore and remove them fairly easily.